optimizer/Transformer/ServerSideRendering.php

```diff
@@ -0,0 +1,546 @@
+        if ($element->hasAttribute('style')) {
+            $styles .= $element->getAttribute('style');
+        }
+        if (! empty($styles)) {
+            $element->setAttribute('style', $styles);
+        }
+
+        $element->setAttribute(self::LAYOUT_ATTRIBUTE, $layout);
+    }
+
+    /**
+     * Get the class to use for a given layout.
+     *
+     * @param string $layout Layout to get the class for.
+     * @return string Class name to use for the layout.
+     */
+    private function getLayoutClass($layout)
+    {
+        if (empty($layout)) {
+            return '';
+        }
+
+        return self::LAYOUT_CLASS_PREFIX . $layout;
+    }
+
+    /**
+     * Add a class to an element.
+     *
+     * This makes sure we keep existing classes on the element.
+     *
+     * @param DOMElement $element Element to add a class to.
+     * @param string     $class   Class to add.
+     */
+    private function addClass(DOMElement $element, $class)
+    {
+        if ($element->hasAttribute('class') && ! empty($element->getAttribute('class'))) {
+            $class = "{$element->getAttribute('class')} {$class}";
+        }
+
+        $element->setAttribute('class', $class);
+    }
+
+    /**
+     * Check whether the provided layout is a layout with a defined size.
+     *
+     * @param string $layout Layout to check.
+     * @return bool Whether the layout has a defined size.
+     */
+    private function isLayoutSizeDefined($layout)
+    {
+        return in_array($layout, Layout::SIZE_DEFINED_LAYOUTS, true);
+    }
+
+    private function maybeAddSizerInto(
+        Document $document,
+        DOMElement $element,
+        $layout,
+        AMP_CSS_Length $width,
+        AMP_CSS_Length $height
+    ) {
+        if ($layout !== Layout::RESPONSIVE
+            || ! $width->is_set()
+            || $width->get_numeral() === 0
+            || ! $height->is_set()
+            || $width->get_unit() !== $height->get_unit()) {
+            return;
+        }
+
+        $padding = $height->get_numeral() / $width->get_numeral() * 100;
+        $sizer   = $document->createElement(self::SIZER_ELEMENT);
+        $sizer->setAttribute('style', sprintf('display:block;padding-top:%1.4F%%;', $padding));
+        $element->insertBefore($sizer, $element->firstChild);
+    }
+
+    /**
+     * Check whether the element has an ancestor of a given tag type.
+     *
+     * @param DOMElement $element Element to check the ancestor tree of.
+     * @param string     $tagName Name of the tag to look for.
+     * @return bool Whether the element has an ancestor of the given tag name.
+     */
+    private function hasAncestorWithTag(DOMElement $element, $tagName)
+    {
+        $parent = $element->parentNode;
+        while ($parent !== null) {
+            if ($parent instanceof DOMElement && $parent->tagName === $tagName) {
+                return true;
+            }
+            $parent = $parent->parentNode;
+        }
+        return false;
+    }
```

Comment:
Since this method is going to be called a lot, perhaps it should be optimized better? When collecting the `amp_elements` can we omit the ones that are inside of `template` elements? Perhaps something to add to the `Document` class? For example:

```php
case 'templateless_amp_elements':
    $this->templateless_amp_elements = $this->xpath->query( ".//*[ starts-with( name(), 'amp-' ) and not( ancestor::template/ancestor::* = current() ) ]", $this->body );
```

Props https://stackoverflow.com/a/32272565

Reply:
As it turns out, `current()` was not yet part of XPath 1.0, which libXML only supports.

![image](https://user-images.githubusercontent.com/83631/71996295-41b48e80-323c-11ea-8c93-abbb8d522d46.png)

Reply:
We can benchmark and optimize this in a subsequent PR.